src/connector/src/sink/pulsar.rs

```diff
@@ -0,0 +1,321 @@
+
+/// The delivery buffer queue size
+/// When the `SendFuture` the current `send_future_buffer`
+/// is buffering is greater than this size, then enforcing commit once
+const PULSAR_SEND_FUTURE_BUFFER_MAX_SIZE: usize = 65536;
+
+const fn _default_max_retries() -> u32 {
+    3
+}
+
+const fn _default_retry_backoff() -> Duration {
+    Duration::from_millis(100)
+}
+
+#[derive(Debug, Clone, Deserialize)]
+pub struct PulsarConfig {
+    #[serde(
+        rename = "properties.retry.max",
+        default = "_default_max_retries",
+        deserialize_with = "deserialize_u32_from_string"
+    )]
+    pub max_retry_num: u32,
+
+    #[serde(
+        rename = "properties.retry.interval",
+        default = "_default_retry_backoff",
+        deserialize_with = "deserialize_duration_from_string"
+    )]
+    pub retry_interval: Duration,
+
+    #[serde(flatten)]
+    pub common: PulsarCommon,
+
+    pub r#type: String, // accept "append-only" or "upsert"
+}
+
+impl PulsarConfig {
+    pub fn from_hashmap(values: HashMap<String, String>) -> Result<Self> {
+        let config = serde_json::from_value::<PulsarConfig>(serde_json::to_value(values).unwrap())
+            .map_err(|e| SinkError::Config(anyhow!(e)))?;
+        
+        if config.r#type != SINK_TYPE_APPEND_ONLY
+            && config.r#type != SINK_TYPE_UPSERT {
+            return Err(SinkError::Config(anyhow!(
+                "`{}` must be {}, or {}",
+                SINK_TYPE_OPTION,
+                SINK_TYPE_APPEND_ONLY,
+                SINK_TYPE_UPSERT
+            )));
+        }
+        Ok(config)
+    }
+}
+
+#[derive(Debug)]
+pub struct PulsarSink {
+    pub config: PulsarConfig,
+    schema: Schema,
+    downstream_pk: Vec<usize>,
+    is_append_only: bool,
+}
+
+impl PulsarSink {
+    pub fn new(config: PulsarConfig, param: SinkParam) -> Self {
+        Self {
+            config,
+            schema: param.schema(),
+            downstream_pk: param.downstream_pk,
+            is_append_only: param.sink_type.is_append_only(),
+        }
+    }
+}
+
+#[async_trait]
+impl Sink for PulsarSink {
+    type Coordinator = DummySinkCommitCoordinator;
+    type Writer = PulsarSinkWriter;
+    async fn new_writer(&self, _writer_param: SinkWriterParam) -> Result<Self::Writer> {
+        PulsarSinkWriter::new(
+            self.config.clone(), 
+            self.schema.clone(), 
+            self.downstream_pk.clone(),
+            self.is_append_only,
+        ).await
+    }
+
+    async fn validate(&self, _client: Option<ConnectorClient>) -> Result<()> {
+        // For upsert Pulsar sink, the primary key must be defined.
+        if !self.is_append_only && self.downstream_pk.is_empty() {
+            return Err(SinkError::Config(anyhow!(
+                "primary key not defined for {} pulsar sink (please define in `primary_key` field)",
+                self.config.r#type
+            )));
+        }
+
+        // TODO: validate pulsar connection
```

Comment:
you can do as the source enumerator does here. but you only need to know whether the call is successful or not
https://github.com/risingwavelabs/risingwave/blob/59bb645b27d6345abddfebef5e3095c56b74ffe7/src/connector/src/source/pulsar/enumerator/client.rs#L80-L108

Reply:
Producer in pulsar-rs has a [check_connection](https://docs.rs/pulsar/latest/pulsar/producer/struct.Producer.html#method.check_connection) method. However, calling it on a valid producer client which can send messages successfully, causes a connection error for no reason. Therefore I just build a producer client to validate the connection. The topic/url/token will be validate when building the producer.